pkg/operator/ceph/csi/spec.go

```diff
@@ -676,12 +676,29 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 	return nil
 }
 
-func (r *ReconcileCSI) stopDrivers(ver *version.Info) error {
+// return false is rook operator is not the owner of csi driver,
+// so csi driver do not get deleted as it isn't deployed by
+// rook operator and someone externally installed it
+func (r *ReconcileCSI) validateDriverOwner(ownerInfo *k8sutil.OwnerInfo, deploymentName string) bool {
+	getDeployment, err := r.context.Clientset.AppsV1().Deployments(r.opConfig.OperatorNamespace).Get(r.opManagerContext, deploymentName, metav1.GetOptions{})
+	if err != nil {
+		logger.Warning("csi deployment pod not present")
+	}
+
+	if getDeployment.OwnerReferences[0].UID != ownerInfo.GetUID() {
+		logger.Debugf("csi owner reference UID %s matches operator owner reference UID %s, so delete the csi driver", getDeployment.OwnerReferences[0].UID, ownerInfo.GetUID())
+		return false
+	}
+
+	return true
+}
+
+func (r *ReconcileCSI) stopDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerInfo) error {
 	RBDDriverName = fmt.Sprintf("%s.rbd.csi.ceph.com", r.opConfig.OperatorNamespace)
 	CephFSDriverName = fmt.Sprintf("%s.cephfs.csi.ceph.com", r.opConfig.OperatorNamespace)
 	NFSDriverName = fmt.Sprintf("%s.nfs.csi.ceph.com", r.opConfig.OperatorNamespace)
 
-	if !EnableRBD {
+	if !EnableRBD && r.validateDriverOwner(ownerInfo, csiRBDProvisioner) {
```

Comment:
This check seems wrong, why we are checking provisioner pod and deleting daemonset?

Reply:
we always create and delete provisioner pod and daemonset back to back, so checking anyone would be fine.
And there isn't any separate flag to create or delete any one of them. 

If wanted we can add a separate check for both, but IMO not needed

Reply:
What about the another case where rook operator deleted the deployment but got restarted before deleting other object like DaemonSet,  csidriver etc?

Reply:
That case make sense, for that we should keep it separate, thanks for the feedback

Reply:
@travisn your comments?

Reply:
So,
1) Check for the provisioner deployment existence
   1.1) If present and owner matches go and delete csi driver
   1.2) if not present go and delete the csi driver
   1.3) if present and owner is not same do not delete  the csi driver